drivers/mspi/mspi_nrfe.c

```diff
@@ -236,75 +162,146 @@ static int nrfe_mspi_wait_for_response(nrfe_mspi_opcode_t opcode, uint32_t timeo
+#else
+	uint32_t len = sizeof(nrfe_mspi_xfer_packet_msg_t) + packet->num_bytes;
+#endif
+	uint8_t buffer[len];
+	nrfe_mspi_xfer_packet_msg_t *xfer_packet = (nrfe_mspi_xfer_packet_msg_t *)buffer;
+
+	xfer_packet->opcode = opcode;
+	xfer_packet->command = packet->cmd;
+	xfer_packet->address = packet->address;
+	xfer_packet->num_bytes = packet->num_bytes;
+
+#ifdef CONFIG_MSPI_NRFE_IPC_NO_COPY
+	/*
+	 * Check for alignment problems. With no-copy mode, FLPR expects data buffer to be aligned
+	 * to word size.
+	 */
+	if (((uint32_t)packet->data_buf) % sizeof(uint32_t) != 0) {
+		memcpy((void *)(buffer + sizeof(nrfe_mspi_xfer_packet_msg_t)),
+		       (void *)packet->data_buf, packet->num_bytes);
+		xfer_packet->data = buffer + sizeof(nrfe_mspi_xfer_packet_msg_t);
+	} else {
+		xfer_packet->data = packet->data_buf;
+	}
+#else
+	memcpy((void *)xfer_packet->data, (void *)packet->data_buf, packet->num_bytes);
+#endif
+
+	/* Wait for the transfer to complete and receive data. */
+	rc = send_and_wait(xfer_packet->opcode, xfer_packet, len);
+
+	if ((packet->dir == MSPI_RX) && (ipc_receive_buffer != NULL) && (ipc_received > 0)) {
+		/*
+		 * It is not possible to check whether received data is valid, so packet->num_bytes
+		 * should always be equal to ipc_received. If it is not, then something went wrong.
+		 */
+		if (packet->num_bytes != ipc_received) {
+			rc = -EIO;
+		} else {
+			memcpy((void *)packet->data_buf, (void *)ipc_receive_buffer, ipc_received);
+		}
+
+		/* Clear the receive buffer pointer and size */
+		ipc_receive_buffer = NULL;
+		ipc_received = 0;
+	}
+
+	return rc;
+}
+
+static int api_transceive(const struct device *dev, const struct mspi_dev_id *dev_id,
+			  const struct mspi_xfer *req)
+{
+	struct mspi_nrfe_data *drv_data = dev->data;
+	uint32_t packet_idx = 0;
+	int rc;
+
+	/* TODO: add support for asynchronous transfers */
+	if (req->async) {
+		return -ENOTSUP;
+	}
+
+	if (req->num_packet == 0 || !req->packets ||
+	    req->timeout > CONFIG_MSPI_COMPLETION_TIMEOUT_TOLERANCE) {
+		return -EFAULT;
+	}
+
+	drv_data->xfer_config_msg.opcode = NRFE_MSPI_CONFIG_XFER;
+	drv_data->xfer_config_msg.xfer_config.device_index = dev_id->dev_idx;
+	drv_data->xfer_config_msg.xfer_config.command_length = req->cmd_length;
+	drv_data->xfer_config_msg.xfer_config.address_length = req->addr_length;
+	drv_data->xfer_config_msg.xfer_config.hold_ce = req->hold_ce;
+	drv_data->xfer_config_msg.xfer_config.tx_dummy = req->tx_dummy;
+	drv_data->xfer_config_msg.xfer_config.rx_dummy = req->rx_dummy;
+
+	rc = send_and_wait(NRFE_MSPI_CONFIG_XFER, (void *)&drv_data->xfer_config_msg,
+		       sizeof(nrfe_mspi_xfer_config_msg_t));
+
+	if (rc < 0) {
+		LOG_ERR("Send xfer config error: %d", rc);
+		return rc;
+	}
+
+	while (packet_idx < req->num_packet) {
+		rc = xfer_packet((struct mspi_xfer *)req, packet_idx);
+		if (rc < 0) {
+			LOG_ERR("xfer_packet error: %d packet_idx: %d", rc, packet_idx);
+			return rc;
+		}
+		packet_idx++;
+	}
```

Comment:
Nit: for loop would be more suitable, since `packet_idx` is not used outside of the loop. But can be left as it is.